src/workers/layouts/nearestParentVertical.ts

```diff
@@ -0,0 +1,444 @@
+import { HorizontalLayout } from '@/workers/layouts/horizontal'
+import { VerticalLayout } from '@/workers/layouts/vertical'
+import { RunGraphNode } from "@/models"
+import { ClientLayoutMessage } from '@/workers/runGraph'
+
+type NodeShoveDirection = 1 | -1
+
+export async function getVerticalNearestParentLayout(message: ClientLayoutMessage, horizontal: HorizontalLayout): Promise<VerticalLayout> {
+  const defaultNearestParentPosition = 0
+  const minimumNodeEdgeGap = 16
+  const nodeShoveRecords = new Map<string, NodeShoveDirection>()
+
+  const layout: VerticalLayout = new Map()
+
+  for (const [nodeId] of message.data.nodes) {
+    const node = message.data.nodes.get(nodeId)
+
+    if (!node) {
+      console.warn('NearestParentLayout: Node was not found in the data', nodeId)
+      continue
+    }
+
+    const row = await getNearestParentPosition(node)
+
+    layout.set(nodeId, row)
+  }
+
+  purgeNegativeLayoutPositions()
+
+  return layout
+
+  async function getNearestParentPosition(node: RunGraphNode): Promise<number> {
+    const nodeStartX = horizontal.get(node.id)
+
+    if (nodeStartX === undefined) {
+      console.warn('NearestParentLayout: Node was not found in the horizontal layout', node.id)
+      return defaultNearestParentPosition
+    }
+
+    // if one dependency
+    if (node.parents && node.parents.length === 1) {
+      if (layout.has(node.parents[0].id)) {
+        return await placeNearUpstreamNode(node.parents[0].id, nodeStartX)
+      }
+
+      console.warn('NearestParentLayout: Parent node not found in layout', node.parents[0].id)
+      return defaultNearestParentPosition
+    }
+
+    // if more than one dependency – add to the middle of upstream dependencies
+    if (node.parents && node.parents.length > 0) {
+      const upstreamRows = node.parents.map(({id}) => {
+        const row = layout.get(id)
+
+        if (row === undefined) {
+          console.warn('NearestParentLayout: Parent node not found in layout', id)
+          return defaultNearestParentPosition
+        }
+
+        return row
+      })
+
+      const upstreamRowsSum = upstreamRows.reduce((sum, position) => sum + position, 0)
+      const upstreamRowsAverage = upstreamRowsSum / upstreamRows.length
+      const row = Math.round(upstreamRowsAverage)
+
+      if (isPositionTaken(nodeStartX, row)) {
+        const overlappingNodes = getOverlappingNodeIds(
+          nodeStartX,
+          row,
+        )!
+
+        const overlappingParentNodes = overlappingNodes.filter(layoutId => {
+          return node.parents?.some(({id}) => id === layoutId)
+        })
```

Comment:
This is a potentially expensive operation because we're calling some within a filter. Maybe we could set the data up differently to make this a one shot calculation